pkgs/development/compilers/go/1.16.nix

```diff
@@ -0,0 +1,260 @@
+{ lib, stdenv, fetchurl, tzdata, iana-etc, runCommand
+, perl, which, pkg-config, patch, procps, pcre, cacert, Security, Foundation
+, mailcap, runtimeShell
+, buildPackages
+, pkgsBuildTarget
+, fetchpatch
+}:
+
+let
+
+  inherit (lib) optionals optionalString;
+
+  goBootstrap = runCommand "go-bootstrap" {} ''
+    mkdir $out
+    cp -rf ${buildPackages.go_bootstrap}/* $out/
+    chmod -R u+w $out
+    find $out -name "*.c" -delete
+    cp -rf $out/bin/* $out/share/go/bin/
+  '';
+
+  goarch = platform: {
+    "i686" = "386";
+    "x86_64" = "amd64";
+    "aarch64" = "arm64";
+    "arm" = "arm";
+    "armv5tel" = "arm";
+    "armv6l" = "arm";
+    "armv7l" = "arm";
+    "powerpc64le" = "ppc64le";
+  }.${platform.parsed.cpu.name} or (throw "Unsupported system");
+
+  # We need a target compiler which is still runnable at build time,
+  # to handle the cross-building case where build != host == target
+  targetCC = pkgsBuildTarget.targetPackages.stdenv.cc;
+in
+
+stdenv.mkDerivation rec {
+  pname = "go";
+  version = "1.16rc1";
+
+  src = fetchurl {
+    url = "https://dl.google.com/go/go${version}.src.tar.gz";
+    sha256 = "1x9zfvjiqd512askzif184xnig0g9wmcr1j0c4qxn88dkngmccva";
+  };
+
+  # perl is used for testing go vet
+  nativeBuildInputs = [ perl which pkg-config patch procps ];
+  buildInputs = [ cacert pcre ]
+    ++ optionals stdenv.isLinux [ stdenv.cc.libc.out ]
+    ++ optionals (stdenv.hostPlatform.libc == "glibc") [ stdenv.cc.libc.static ];
+
+  depsTargetTargetPropagated = optionals stdenv.isDarwin [ Security Foundation ];
```

Comment:
I think you also need the following for 1.16 (plus `xcbuild` in inputs):

```
propagatedBuildInputs = optionals stdenv.isDarwin [ xcbuild ];
```

See: https://github.com/golang/go/commit/7e25bdba5e79d223566745fca2410f725a6dedda

Otherwise you won't be able to link packages inside of Nix build environment:

```
/nix/store/61kpq6gcyp94j87ni6i12vl4vpmsgb02-go-1.16rc1/share/go/pkg/tool/darwin_amd64/link: /nix/store/61kpq6gcyp94j87ni6i12vl4vpmsgb02-go-1.16rc1/share/go/pkg/tool/darwin_amd64/link: running dsymutil failed: exec: "xcrun": executable file not found in $PATH
```

Go will still work outside of Nix if you run `go`, because it will find `/usr/bin/xcrun` from developer tools.